File: sdk/verifiers/ts/src/aarp/canonical.ts
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

// RFC 8785 JCS canonicalization, ported to match Go's
// internal/contract.Canonicalize byte-for-byte:
//
//   - strings are NFC-normalized (String.prototype.normalize("NFC")),
//   - object keys are sorted ascending by Unicode code point,
//   - output is compact (no insignificant whitespace),
//   - the bytes < > & and U+2028/U+2029 are escaped the way Go's
//     encoding/json does by default,
//   - floats are rejected; only integer numbers are emitted.
//
// This is the single canonicalizer used both to recompute the signed payload
// digest and to emit the comparable appraisal output, exactly as the Go
// reference uses one Canonicalize for both surfaces.

// compareCodePointStrings orders two strings ascending by Unicode code point,
// matching Go's lexicographic comparison of the NFC-normalized UTF-8 key bytes.
// Iterating by code point (Array.from) rather than UTF-16 code unit keeps
// supplementary-plane keys ordered the way Go orders their UTF-8 bytes.
export function compareCodePointStrings(a: string, b: string): number {
  const left = Array.from(a);
  const right = Array.from(b);
  const n = Math.min(left.length, right.length);
  for (let i = 0; i < n; i++) {
    const leftCodePoint = left[i]?.codePointAt(0) ?? 0;
    const rightCodePoint = right[i]?.codePointAt(0) ?? 0;
    if (leftCodePoint !== rightCodePoint) return leftCodePoint - rightCodePoint;
  }
  return left.length - right.length;
}

// goHTMLEscape applies the same default escaping Go's encoding/json does to a
// JSON-serialized string or document: < > & become their \u00xx forms and the
// JSON-unsafe line/paragraph separators U+2028/U+2029 become \u2028/\u2029.
// JSON.stringify and Go's encoding/json already agree on quote, backslash, and
// C0 control escaping, so only these five bytes need fixing up.
function goHTMLEscape(serialized: string): string {
  return serialized
    .replace(/</g, "\\u003c")
    .replace(/>/g, "\\u003e")
    .replace(/&/g, "\\u0026")
    .replace(/\u2028/gu, "\\u2028")
    .replace(/\u2029/gu, "\\u2029");
}

// replaceSurrogates matches Go encoding/json's handling of UNPAIRED UTF-16
// surrogate code units: a lone high or low surrogate becomes U+FFFD before
// normalization and JSON string escaping. A VALID surrogate pair (high followed
// by low) is a legitimate astral code point (e.g. an emoji) and MUST be
// preserved intact -- naively replacing every surrogate code unit would mangle
// astral characters to two U+FFFD, computing different canonical bytes than Go,
// Rust, and Python, which is exactly the cross-language divergence this corpus
// exists to prevent. The strict parser already substitutes lone surrogates at
// decode time, so this is a defense-in-depth pass that must not corrupt pairs.
function replaceSurrogates(s: string): string {
  let out = "";
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = s.charCodeAt(i + 1);
      if (next >= 0xdc00 && next <= 0xdfff) {
        out += s[i]! + s[i + 1]!; // valid astral pair: keep both code units
        i++;
        continue;
      }
      out += "\uFFFD"; // lone high surrogate
      continue;
    }
    if (code >= 0xdc00 && code <= 0xdfff) {
      out += "\uFFFD"; // lone low surrogate
      continue;
    }
    out += s[i] ?? "";
  }
  return out;
}

// CanonicalizeError mirrors the Go canonicalizer's fatal conditions: a float (or
// non-integer number), an unsupported value type, or an NFC key collision.
export class CanonicalizeError extends Error {
  readonly code = 1;
}

// encodeString NFC-normalizes a string and emits its Go-encoding/json escaping.
function encodeString(s: string): string {
  return goHTMLEscape(JSON.stringify(replaceSurrogates(s).normalize("NFC")));
}

// isPlainObject narrows to a non-array, non-null object.
function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

// canonicalizeValue serializes a single JSON value into the buffer of pieces.
// Numbers are validated as safe integers earlier (enforceSafeNumbers) for parsed
// envelopes; here we still reject any float we are handed so the comparable-output
// path (which builds objects directly) can never emit a non-integer number.
function canonicalizeValue(v: unknown, out: string[]): void {
  if (v === null || v === undefined) {
    out.push("null");
    return;
  }
  if (typeof v === "boolean") {
    out.push(v ? "true" : "false");
    return;
  }
  if (typeof v === "string") {
    out.push(encodeString(v));
    return;
  }
  if (typeof v === "bigint") {
    out.push(v.toString());
    return;
  }
  if (typeof v === "number") {
    if (!Number.isInteger(v)) {
      throw new CanonicalizeError(`float not allowed in canonicalization: ${v}`);
    }
    out.push(String(v));
    return;
  }
  if (Array.isArray(v)) {
    out.push("[");
    for (let i = 0; i < v.length; i++) {
      if (i > 0) out.push(",");
      canonicalizeValue(v[i], out);
    }
    out.push("]");
    return;
  }
  if (isPlainObject(v)) {
    canonicalizeObject(v, out);
    return;
  }
  throw new CanonicalizeError(`unsupported type in canonicalization: ${typeof v}`);
}

// canonicalizeObject sorts keys by NFC code point and rejects NFC collisions,
// mirroring the Go canonicalizer's duplicate-after-normalization guard.
function canonicalizeObject(obj: Record<string, unknown>, out: string[]): void {
  const pairs = Object.keys(obj).map((orig) => ({
    nfc: replaceSurrogates(orig).normalize("NFC"),
    orig,
  }));
  pairs.sort((a, b) => compareCodePointStrings(a.nfc, b.nfc));
  for (let i = 1; i < pairs.length; i++) {
    const cur = pairs[i];
    const prev = pairs[i - 1];
    if (cur && prev && cur.nfc === prev.nfc && cur.orig !== prev.orig) {
      throw new CanonicalizeError(`NFC collision between ${prev.orig} and ${cur.orig}`);
    }
  }
  out.push("{");
  for (let i = 0; i < pairs.length; i++) {
    const p = pairs[i];
    if (!p) continue;
    if (i > 0) out.push(",");
    out.push(goHTMLEscape(JSON.stringify(p.nfc)));
    out.push(":");
    canonicalizeValue(obj[p.orig], out);
  }
  out.push("}");
}

// canonicalize returns the JCS-canonical UTF-8 string for the given value.
export function canonicalize(v: unknown): string {
  const out: string[] = [];
  canonicalizeValue(v, out);
  return out.join("");
}

// canonicalizeBytes returns the JCS-canonical bytes for the given value.
export function canonicalizeBytes(v: unknown): Buffer {
  return Buffer.from(canonicalize(v), "utf8");
}

File: sdk/verifiers/python/src/pipelock_aarp_verify/canonical.py
# Copyright 2026 Josh Waldrep
# SPDX-License-Identifier: Apache-2.0

"""RFC 8785 JCS canonicalization, ported byte-for-byte from the Go reference.

The reference is ``internal/contract/canonicalize.go``. Two verifiers in two
languages that canonicalize the same value MUST produce byte-identical output;
any divergence is the bug class the AARP corpus exists to kill. The subtle parts
this module reproduces exactly:

  - Strings are NFC-normalized (``unicodedata.normalize("NFC", s)``).
  - Object keys are sorted ascending by Unicode code point (Python's default
    ``str`` ordering is by code point, which is correct).
  - Compact output: no whitespace, ``,`` and ``:`` separators only.
  - Per-string escaping matches Go ``json.Marshal``: ``<`` ``>`` ``&`` are
    escaped to ``\\u003c`` ``\\u003e`` ``\\u0026``; U+2028 and U+2029 become
    ``\\u2028`` ``\\u2029``; control characters use the Go short forms where Go
    uses them (``\\b \\t \\n \\f \\r``) and ``\\u00xx`` otherwise; all other
    code points (including DEL U+007F and any non-ASCII) are emitted raw UTF-8.
  - Floats are rejected; integers are emitted verbatim from their source text.
  - NFC key collisions are rejected as duplicate keys.
"""

from __future__ import annotations

import unicodedata
from typing import Any

from .number import IJSONNumber


class CanonicalizeError(Exception):
    """A value could not be canonicalized (float, bad type, or key collision)."""


# Short escape forms Go's encoder emits for these control characters. Every
# other control character below U+0020 is emitted as ``\u00xx``.
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
}


def _replace_surrogates(s: str) -> str:
    """Match Go encoding/json: unpaired UTF-16 surrogates decode as U+FFFD."""
    return "".join("\ufffd" if 0xD800 <= ord(ch) <= 0xDFFF else ch for ch in s)


def _escape_string(s: str) -> str:
    """Return the JSON string literal for ``s`` exactly as Go ``json.Marshal``.

    ``s`` is NFC-normalized first, matching the Go canonicalizer which calls
    ``norm.NFC.String`` on every string and map key.
    """
    s = unicodedata.normalize("NFC", _replace_surrogates(s))
    out = ['"']
    for ch in s:
        code = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif code in _SHORT_ESCAPES:
            out.append(_SHORT_ESCAPES[code])
        elif code < 0x20:
            out.append(f"\\u{code:04x}")
        elif ch == "<":
            out.append("\\u003c")
        elif ch == ">":
            out.append("\\u003e")
        elif ch == "&":
            out.append("\\u0026")
        elif code == 0x2028:
            out.append("\\u2028")
        elif code == 0x2029:
            out.append("\\u2029")
        else:
            # All other code points (DEL U+007F included) emit raw UTF-8.
            out.append(ch)
    out.append('"')
    return "".join(out)


def _canon_into(parts: list[str], v: Any) -> None:
    if v is None:
        parts.append("null")
        return
    if v is True:
        parts.append("true")
        return
    if v is False:
        parts.append("false")
        return
    if isinstance(v, str):
        parts.append(_escape_string(v))
        return
    if isinstance(v, IJSONNumber):
        # Source-text integer literal, already validated as a safe integer by the
        # number-safety pass. Emit verbatim so the canonical bytes are stable.
        parts.append(v.literal)
        return
    if isinstance(v, bool):  # pragma: no cover - handled above, kept for clarity
        parts.append("true" if v else "false")
        return
    if isinstance(v, int):
        parts.append(str(v))
        return
    if isinstance(v, float):
        raise CanonicalizeError(
            "float not allowed in canonicalization; use decimal string"
        )
    if isinstance(v, (list, tuple)):
        parts.append("[")
        for i, item in enumerate(v):
            if i > 0:
                parts.append(",")
            _canon_into(parts, item)
        parts.append("]")
        return
    if isinstance(v, dict):
        # Track NFC-normalized key (for sort + output) alongside the original
        # (for value lookup), mirroring the Go canonicalizer.
        pairs = []
        for k in v:
            if not isinstance(k, str):
                raise CanonicalizeError(f"non-string map key: {k!r}")
            pairs.append((unicodedata.normalize("NFC", k), k))
        pairs.sort(key=lambda p: p[0])
        for i in range(1, len(pairs)):
            if pairs[i][0] == pairs[i - 1][0] and pairs[i][1] != pairs[i - 1][1]:
                raise CanonicalizeError(
                    f"duplicate key in JSON object: NFC collision between "
                    f"{pairs[i - 1][1]!r} and {pairs[i][1]!r}"
                )
        parts.append("{")
        for i, (nfc_key, orig_key) in enumerate(pairs):
            if i > 0:
                parts.append(",")
            parts.append(_escape_string(nfc_key))
            parts.append(":")
            _canon_into(parts, v[orig_key])
        parts.append("}")
        return
    raise CanonicalizeError(f"canonicalize: unsupported type {type(v).__name__}")


def canonicalize(v: Any) -> bytes:
    """Return the RFC 8785 JCS canonical bytes for ``v`` (UTF-8 encoded)."""
    parts: list[str] = []
    _canon_into(parts, v)
    return "".join(parts).encode("utf-8")
